File: code/packages/java/hash-map/src/main/java/com/codingadventures/hashmap/HashMap.java
package com.codingadventures.hashmap;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public final class HashMap<K, V> {
    private final LinkedHashMap<K, V> delegate;

    public HashMap() {
        this.delegate = new LinkedHashMap<>();
    }

    public HashMap(Map<K, V> values) {
        this.delegate = new LinkedHashMap<>(values);
    }

    public int size() {
        return delegate.size();
    }

    public boolean isEmpty() {
        return delegate.isEmpty();
    }

    public boolean has(K key) {
        return delegate.containsKey(key);
    }

    public V get(K key) {
        return delegate.get(key);
    }

    public V getOrDefault(K key, V defaultValue) {
        return delegate.getOrDefault(key, defaultValue);
    }

    public HashMap<K, V> set(K key, V value) {
        delegate.put(key, value);
        return this;
    }

    public boolean delete(K key) {
        return delegate.remove(key) != null;
    }

    public void clear() {
        delegate.clear();
    }

    public List<K> keys() {
        return new ArrayList<>(delegate.keySet());
    }

    public List<V> values() {
        return new ArrayList<>(delegate.values());
    }

    public List<Map.Entry<K, V>> entries() {
        List<Map.Entry<K, V>> entries = new ArrayList<>();
        for (Map.Entry<K, V> entry : delegate.entrySet()) {
            entries.add(Map.entry(entry.getKey(), entry.getValue()));
        }
        return entries;
    }

    public HashMap<K, V> copy() {
        return new HashMap<>(delegate);
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof HashMap<?, ?> that)) {
            return false;
        }
        return Objects.equals(delegate, that.delegate);
    }

    @Override
    public int hashCode() {
        return delegate.hashCode();
    }

    @Override
    public String toString() {
        return delegate.toString();
    }
}

File: code/packages/kotlin/hash-map/src/main/kotlin/com/codingadventures/hashmap/HashMap.kt
package com.codingadventures.hashmap

class HashMap<K, V>(values: Map<K, V> = emptyMap()) {
    private val delegate = linkedMapOf<K, V>().apply { putAll(values) }

    val size: Int get() = delegate.size
    fun isEmpty(): Boolean = delegate.isEmpty()
    fun has(key: K): Boolean = delegate.containsKey(key)
    operator fun get(key: K): V? = delegate[key]
    fun getOrDefault(key: K, defaultValue: V): V = delegate.getOrDefault(key, defaultValue)
    fun set(key: K, value: V): HashMap<K, V> {
        delegate[key] = value
        return this
    }
    fun delete(key: K): Boolean = delegate.remove(key) != null
    fun clear() = delegate.clear()
    fun keys(): List<K> = delegate.keys.toList()
    fun values(): List<V> = delegate.values.toList()
    fun entries(): List<Pair<K, V>> = delegate.entries.map { it.key to it.value }
    fun copy(): HashMap<K, V> = HashMap(delegate)

    override fun equals(other: Any?): Boolean = other is HashMap<*, *> && delegate == other.delegate
    override fun hashCode(): Int = delegate.hashCode()
    override fun toString(): String = delegate.toString()
}
